File: core-util/src/main/java/org/signal/core/util/logging/Log.java
package org.signal.core.util.logging;

import android.annotation.SuppressLint;

import androidx.annotation.MainThread;
import androidx.annotation.NonNull;

@SuppressLint("LogNotSignal")
public final class Log {

  private static final Logger NOOP_LOGGER = new NoopLogger();

  private static InternalCheck internalCheck;
  private static Logger        logger = new AndroidLogger();

  /**
   * @param internalCheck A checker that will indicate if this is an internal user
   * @param loggers A list of loggers that will be given every log statement.
   */
  @MainThread
  public static void initialize(@NonNull InternalCheck internalCheck, Logger... loggers) {
    Log.internalCheck = internalCheck;
    Log.logger        = new CompoundLogger(loggers);
  }

  public static void initialize(Logger... loggers) {
    initialize(() -> false, loggers);
  }

  public static void v(String tag, String message) {
    v(tag, message, null);
  }

  public static void d(String tag, String message) {
    d(tag, message, null);
  }

  public static void i(String tag, String message) {
    i(tag, message, null);
  }

  public static void w(String tag, String message) {
    w(tag, message, null);
  }

  public static void e(String tag, String message) {
    e(tag, message, null);
  }

  public static void v(String tag, Throwable t) {
    v(tag, null, t);
  }

  public static void d(String tag, Throwable t) {
    d(tag, null, t);
  }

  public static void i(String tag, Throwable t) {
    i(tag, null, t);
  }

  public static void w(String tag, Throwable t) {
    w(tag, null, t);
  }

  public static void e(String tag, Throwable t) {
    e(tag, null, t);
  }

  public static void v(String tag, String message, Throwable t) {
    logger.v(tag, message, t);
  }

  public static void d(String tag, String message, Throwable t) {
    logger.d(tag, message, t);
  }

  public static void i(String tag, String message, Throwable t) {
    logger.i(tag, message, t);
  }

  public static void w(String tag, String message, Throwable t) {
    logger.w(tag, message, t);
  }

  public static void e(String tag, String message, Throwable t) {
    logger.e(tag, message, t);
  }

  public static void v(String tag, String message, boolean keepLonger) {
    logger.v(tag, message, keepLonger);
  }

  public static void d(String tag, String message, boolean keepLonger) {
    logger.d(tag, message, keepLonger);
  }

  public static void i(String tag, String message, boolean keepLonger) {
    logger.i(tag, message, keepLonger);
  }

  public static void w(String tag, String message, boolean keepLonger) {
    logger.w(tag, message, keepLonger);
  }

  public static void e(String tag, String message, boolean keepLonger) {
    logger.e(tag, message, keepLonger);
  }

  public static void v(String tag, String message, Throwable t, boolean keepLonger) {
    logger.v(tag, message, t, keepLonger);
  }

  public static void d(String tag, String message, Throwable t, boolean keepLonger) {
    logger.d(tag, message, t, keepLonger);
  }

  public static void i(String tag, String message, Throwable t, boolean keepLonger) {
    logger.i(tag, message, t, keepLonger);
  }

  public static void w(String tag, String message, Throwable t, boolean keepLonger) {
    logger.w(tag, message, t, keepLonger);
  }

  public static void e(String tag, String message, Throwable t, boolean keepLonger) {
    logger.e(tag, message, t, keepLonger);
  }

  public static @NonNull String tag(Class<?> clazz) {
    String simpleName = clazz.getSimpleName();
    if (simpleName.length() > 23) {
      return simpleName.substring(0, 23);
    }
    return simpleName;
  }

  /**
   * Important: This is not something that can be used to log PII. Instead, it's intended use is for
   * logs that might be too verbose or otherwise unnecessary for public users.
   *
   * @return The normal logger if this is an internal user, or a no-op logger if it isn't.
   */
  public static Logger internal() {
    if (internalCheck.isInternal()) {
      return logger;
    } else {
      return NOOP_LOGGER;
    }
  }

  public static void blockUntilAllWritesFinished() {
    logger.flush();
  }

  public static abstract class Logger {

    public abstract void v(String tag, String message, Throwable t, boolean keepLonger);
    public abstract void d(String tag, String message, Throwable t, boolean keepLonger);
    public abstract void i(String tag, String message, Throwable t, boolean keepLonger);
    public abstract void w(String tag, String message, Throwable t, boolean keepLonger);
    public abstract void e(String tag, String message, Throwable t, boolean keepLonger);
    public abstract void flush();

    public void v(String tag, String message, boolean keepLonger) {
      v(tag, message, null, keepLonger);
    }

    public void d(String tag, String message, boolean keepLonger) {
      d(tag, message, null, keepLonger);
    }

    public void i(String tag, String message, boolean keepLonger) {
      i(tag, message, null, keepLonger);
    }

    public void w(String tag, String message, boolean keepLonger) {
      w(tag, message, null, keepLonger);
    }

    public void e(String tag, String message, boolean keepLonger) {
      e(tag, message, null, keepLonger);
    }

    public void v(String tag, String message, Throwable t) {
      v(tag, message, t, false);
    }

    public void d(String tag, String message, Throwable t) {
      d(tag, message, t, false);
    }

    public void i(String tag, String message, Throwable t) {
      i(tag, message, t, false);
    }

    public void w(String tag, String message, Throwable t) {
      w(tag, message, t, false);
    }

    public void e(String tag, String message, Throwable t) {
      e(tag, message, t, false);
    }

    public void v(String tag, String message) {
      v(tag, message, null);
    }

    public void d(String tag, String message) {
      d(tag, message, null);
    }

    public void i(String tag, String message) {
      i(tag, message, null);
    }

    public void w(String tag, String message) {
      w(tag, message, null);
    }

    public void e(String tag, String message) {
      e(tag, message, null);
    }
  }

  public interface InternalCheck {
    boolean isInternal();
  }
}

File: core/util-jvm/src/main/java/org/signal/core/util/logging/Log.kt
/*
 * Copyright 2023 Signal Messenger, LLC
 * SPDX-License-Identifier: AGPL-3.0-only
 */

package org.signal.core.util.logging

import kotlin.reflect.KClass

object Log {
  private val NOOP_LOGGER: Logger = NoopLogger()
  private var internalCheck: InternalCheck? = null
  private var logger: Logger = NoopLogger()

  /**
   * @param internalCheck A checker that will indicate if this is an internal user
   * @param loggers A list of loggers that will be given every log statement.
   */
  @JvmStatic
  fun initialize(internalCheck: InternalCheck?, vararg loggers: Logger) {
    Log.internalCheck = internalCheck
    logger = CompoundLogger(loggers.toList())
  }

  @JvmStatic
  fun initialize(vararg loggers: Logger) {
    initialize({ false }, *loggers)
  }

  @JvmStatic
  fun v(tag: String, message: String) = v(tag, message, null)

  @JvmStatic
  fun v(tag: String, t: Throwable?) = v(tag, null, t)

  @JvmStatic
  fun v(tag: String, message: String?, t: Throwable?) = v(tag, message, t, false)

  @JvmStatic
  fun v(tag: String, message: String?, keepLonger: Boolean) = v(tag, message, null, keepLonger)

  @JvmStatic
  fun v(tag: String, message: String?, t: Throwable?, keepLonger: Boolean) = logger.v(tag, message, t, keepLonger)

  @JvmStatic
  fun d(tag: String, message: String) = d(tag, message, null)

  @JvmStatic
  fun d(tag: String, t: Throwable?) = d(tag, null, t)

  @JvmStatic
  fun d(tag: String, message: String?, t: Throwable? = null) = d(tag, message, t, false)

  @JvmStatic
  fun d(tag: String, message: String?, keepLonger: Boolean) = d(tag, message, null, keepLonger)

  @JvmStatic
  fun d(tag: String, message: String?, t: Throwable?, keepLonger: Boolean) = logger.d(tag, message, t, keepLonger)

  @JvmStatic
  fun i(tag: String, message: String) = i(tag, message, null)

  @JvmStatic
  fun i(tag: String, t: Throwable?) = i(tag, null, t)

  @JvmStatic
  fun i(tag: String, message: String?, t: Throwable? = null) = i(tag, message, t, false)

  @JvmStatic
  fun i(tag: String, message: String?, keepLonger: Boolean) = i(tag, message, null, keepLonger)

  @JvmStatic
  fun i(tag: String, message: String?, t: Throwable?, keepLonger: Boolean) = logger.i(tag, message, t, keepLonger)

  @JvmStatic
  fun w(tag: String, message: String) = w(tag, message, null)

  @JvmStatic
  fun w(tag: String, t: Throwable?) = w(tag, null, t)

  @JvmStatic
  fun w(tag: String, message: String?, t: Throwable? = null) = w(tag, message, t, false)

  @JvmStatic
  fun w(tag: String, message: String?, keepLonger: Boolean) = logger.w(tag, message, keepLonger)

  @JvmStatic
  fun w(tag: String, message: String?, t: Throwable?, keepLonger: Boolean) {
    logger.w(tag, message, t, keepLonger)
  }

  @JvmStatic
  fun e(tag: String, message: String) = e(tag, message, null)

  @JvmStatic
  fun e(tag: String, t: Throwable?) = e(tag, null, t)

  @JvmStatic
  fun e(tag: String, message: String?, t: Throwable? = null) = e(tag, message, t, false)

  @JvmStatic
  fun e(tag: String, message: String?, keepLonger: Boolean) = e(tag, message, null, keepLonger)

  @JvmStatic
  fun e(tag: String, message: String?, t: Throwable?, keepLonger: Boolean) = logger.e(tag, message, t, keepLonger)

  @JvmStatic
  fun tag(clazz: KClass<*>): String {
    return tag(clazz.java)
  }

  @JvmStatic
  fun tag(clazz: Class<*>): String {
    val simpleName = clazz.simpleName

    return if (simpleName.length > 23) {
      simpleName.substring(0, 23)
    } else {
      simpleName
    }
  }

  /**
   * Important: This is not something that can be used to log PII. Instead, it's intended use is for
   * logs that might be too verbose or otherwise unnecessary for public users.
   *
   * @return The normal logger if this is an internal user, or a no-op logger if it isn't.
   */
  @JvmStatic
  fun internal(): Logger {
    return if (internalCheck!!.isInternal()) {
      logger
    } else {
      NOOP_LOGGER
    }
  }

  @JvmStatic
  fun blockUntilAllWritesFinished() {
    logger.flush()
  }

  abstract class Logger {
    abstract fun v(tag: String, message: String?, t: Throwable?, keepLonger: Boolean)
    abstract fun d(tag: String, message: String?, t: Throwable?, keepLonger: Boolean)
    abstract fun i(tag: String, message: String?, t: Throwable?, keepLonger: Boolean)
    abstract fun w(tag: String, message: String?, t: Throwable?, keepLonger: Boolean)
    abstract fun e(tag: String, message: String?, t: Throwable?, keepLonger: Boolean)
    abstract fun flush()

    fun v(tag: String, message: String?) = v(tag, message, null)
    fun v(tag: String, message: String?, t: Throwable?) = v(tag, message, t, false)
    fun v(tag: String, message: String?, keepLonger: Boolean) = v(tag, message, null, keepLonger)

    fun d(tag: String, message: String?) = d(tag, message, null)
    fun d(tag: String, message: String?, t: Throwable?) = d(tag, message, t, false)
    fun d(tag: String, message: String?, keepLonger: Boolean) = d(tag, message, null, keepLonger)

    fun i(tag: String, message: String?) = i(tag, message, null)
    fun i(tag: String, message: String?, t: Throwable?) = i(tag, message, t, false)
    fun i(tag: String, message: String?, keepLonger: Boolean) = i(tag, message, null, keepLonger)

    fun w(tag: String, message: String?) = w(tag, message, null)
    fun w(tag: String, message: String?, t: Throwable?) = w(tag, message, t, false)
    fun w(tag: String, message: String?, keepLonger: Boolean) = w(tag, message, null, keepLonger)

    fun e(tag: String, message: String?) = e(tag, message, null)
    fun e(tag: String, message: String?, t: Throwable?) = e(tag, message, t, false)
    fun e(tag: String, message: String?, keepLonger: Boolean) = e(tag, message, null, keepLonger)
  }

  fun interface InternalCheck {
    fun isInternal(): Boolean
  }
}
